### DISCUSSION TITLE
Make enum.ToString() use Display attribute if present

### DISCUSSION THREAD

**Discussion by fanoI** [dotnet/csharplang#1280] [General]
The enum toString() method is not a lot of useful as it returns the member name instead of a nice description, usually developer use Display(Name = "xxx")] attribute to overcome this limitation and then are forced to write a fake "toString()" method (using reflection), why this cannot be done inside Enum.toString() method itself?

```cs
public enum UserPromotion
{
    None = 0,

    [Display(Name = "Send Job Offers By Mail")]
    SendJobOffersByMail,

    [Display(Name = "Send Job Offers By Sms")]
    SendJobOffersBySms,

    [Display(Name = "Send Other Stuff By Sms")]
    SendPromotionalBySms ,

    [Display(Name = "Send Other Stuff By Mail")]
    SendPromotionalByMail
}
```

For example:

```
var val = UserPromotion.SendPromotionalByMail
Console.WriteLine($"val is {val}"; // val is Send Other Stuff By Mail
```

**Comment by yaakov-h** [comment]
This sounds more like a BCL request than a language request.

This would also be tricky... `DisplayAttribute` is part of `System.ComponentModel.DataAnnotations` or `System.ComponentModel.Annotations` depending on your platform, but `Enum.ToString()` is part of `mscorlib` / `System.Runtime` / etc.

**Comment by fanoI** [comment]
Well the name of the attribute could be not important (a new one defined in mscorlib / System.Runtime could be OK too), the important thing is that should be a way to do this without the need to it "manually" anytime.

**Comment by Logerfo** [comment]
I think you can use a [TypeConverter](https://msdn.microsoft.com/en-us/library/ayybcxe5.aspx?f=255&MSPPError=-2147217396) to achieve that.

**Comment by theunrepentantgeek** [comment]
The `ToString()` method on enumeration values is used a lot - by ORMs using SQL Server, by NoSQL databases, for JSON/YAML/XML serialization, and so on. Changing its behaviour would be a major breaking change.

There's also the small issue of internationalization - hard coding the display strings like this might be suitable for internal software that's never used outside your organization, but for anyone who needs to deal with multiple cultures, this would be a non-starter.

**Comment by MkazemAkhgary** [comment]
you should check this out 

https://stackoverflow.com/questions/479410/enum-tostring-with-user-friendly-strings

**Comment by MkazemAkhgary** [comment]
if you are concerned about performance, use a dictionary like `Dictionary<UserPromotion
, string>` and then simply write `desc[val]` to get friendly name in efficient way.

you can use string array instead of dictionary. and put strings corresponding to index of each enum member. although you should be careful to not miss any member.

---

but `Console.WriteLine` is definitely not the part you want to optimize. so its ok to use reflection for simplicity. because who cares about few milliseconds you are going to save for them?
